iree/compiler/Dialect/Stream/Transforms/PropagateSubviews.cpp

```diff
@@ -0,0 +1,597 @@
+static void expandReturnOp(mlir::ReturnOp op, IndexSet &indexSet,
+                           SubviewMap &subviewMap) {
+  if (!usesResources(op)) return;
+  OpBuilder builder(op);
+  auto operands =
+      expandOperands(op.getLoc(), op.operands(), subviewMap, indexSet, builder);
+  builder.create<mlir::ReturnOp>(op.getLoc(), operands);
+  op.erase();
+}
+
+// Moves subviews across branches.
+// Requires that the ExpandFuncOp pattern handles modifying the block args.
+//
+// Example:
+//    %1 = stream.resource.subview %0[%o] : {%sz} -> {%l}
+//    br ^bb1(%1)
+//  ^bb1(%b):
+//  ->
+//    br ^bb1(%0, %sz, %o, %l)
+//  ^bb1(%a, %b, %c, %d):
+//    %1 = stream.resource.subview %a[%b] : {%c} -> {%d}
+static void expandBranchOp(mlir::BranchOp op, IndexSet &indexSet,
+                           SubviewMap &subviewMap) {
+  OpBuilder builder(op);
+  auto operands = expandOperands(op.getLoc(), op.destOperands(), subviewMap,
+                                 indexSet, builder);
+  builder.create<mlir::BranchOp>(op.getLoc(), op.dest(), operands);
+  op.erase();
+}
+
+static void expandCondBranchOp(mlir::CondBranchOp op, IndexSet &indexSet,
+                               SubviewMap &subviewMap) {
+  if (!usesResources(op)) return;
+  OpBuilder builder(op);
+  builder.create<mlir::CondBranchOp>(
+      op.getLoc(), op.condition(), op.trueDest(),
+      expandOperands(op.getLoc(), op.trueDestOperands(), subviewMap, indexSet,
+                     builder),
+      op.falseDest(),
+      expandOperands(op.getLoc(), op.falseDestOperands(), subviewMap, indexSet,
+                     builder));
+  op.erase();
+}
+
+// Recursively expands resources into (resource, size, offset, length) in |op|.
+static void expandSubviews(Operation *op, ExpandedGlobalMap &globalMap,
+                           IndexSet &indexSet, SubviewMap &subviewMap) {
+  if (auto loadOp = dyn_cast<IREE::Util::GlobalLoadOp>(op)) {
+    expandGlobalLoadOp(loadOp, globalMap, indexSet, subviewMap);
+  } else if (auto storeOp = dyn_cast<IREE::Util::GlobalStoreOp>(op)) {
+    expandGlobalStoreOp(storeOp, globalMap, indexSet, subviewMap);
+  } else if (auto initializerOp = dyn_cast<IREE::Util::InitializerOp>(op)) {
+    expandInitializerOp(initializerOp, globalMap, indexSet, subviewMap);
+  } else if (auto funcOp = dyn_cast<mlir::FuncOp>(op)) {
+    expandFuncOp(funcOp, globalMap, indexSet, subviewMap);
+  } else if (auto callOp = dyn_cast<mlir::CallOp>(op)) {
+    expandCallOp(callOp, indexSet, subviewMap);
+  } else if (auto returnOp = dyn_cast<mlir::ReturnOp>(op)) {
+    expandReturnOp(returnOp, indexSet, subviewMap);
+  } else if (auto branchOp = dyn_cast<mlir::BranchOp>(op)) {
+    expandBranchOp(branchOp, indexSet, subviewMap);
+  } else if (auto condBranchOp = dyn_cast<mlir::CondBranchOp>(op)) {
+    expandCondBranchOp(condBranchOp, indexSet, subviewMap);
+  }
+}
+
+//===----------------------------------------------------------------------===//
+// -iree-stream-propagate-timepoints
```

Comment:
copy/paste

```suggestion
// -iree-stream-propagate-subviews
```

Reply:
shh you'll reveal my secrets to productivity ;P